File: 01/part_2.py
import os.path as path
import re

data_file = path.join("01", "data")

nums = [
    "one",
    "two",
    "three",
    "four",
    "five",
    "six",
    "seven",
    "eight",
    "nine",
    r"\d"
]

nums_dict = {val: i + 1 for i, val in enumerate(nums[:-1])}
nums_dict.update({str(i + 1): i + 1 for i in range(9)})

# Learned: Using a lookahead assertion here simplifies greatly
num_reg_str = "(?=(" + "|".join(nums) + "))"
num_reg = re.compile(num_reg_str)

def main():
    with open(data_file, "r") as f:
        raw_data = f.read().splitlines()
    
    matches = [[(mat.start(), mat.group(1)) for mat in num_reg.finditer(string)] for string in raw_data]
    digits = [(10 * int(nums_dict[mat[0][1]])) + int(nums_dict[mat[-1][1]]) for mat in matches]

    print(sum(digits))
    # 54980

if __name__ == "__main__":
    main()

File: 01/part_2.R
# === Load data ==============================================
data_pull <- function(path = "01/data") {
  con <- file(path, "r")
  on.exit(close(con))

  out <- readLines(con)

  return(out)
}

raw_data <- data_pull()

# === Lookup table ===========================================
numbers <- c(
  "one",
  "two",
  "three",
  "four",
  "five",
  "six",
  "seven",
  "eight",
  "nine"
)

reg <- sprintf("(?=(%s|\\d))", paste(numbers, collapse = "|"))

numbers_lk <- setNames(as.character(1:9), numbers)
numbers_lk <- c(numbers_lk, setNames(1:9, as.character(1:9)))

# === function ================================
get_number <- function(pattern, strs) {
  out <- integer(length(strs))

  all_matches <- regmatches(strs, gregexec(pattern, strs, perl = TRUE))

  for (i in seq_along(all_matches)) {
    # extract numbers
    nums <- all_matches[[i]][2, c(1, ncol(all_matches[[i]]))]
    nums <- as.integer(numbers_lk[nums])

    out[i] <- (10 * nums[1]) + nums[2]
  }

  return(out)
}

# === check =================
clean_data <- get_number(reg, raw_data)

sum(clean_data)
# 54980
